Add feature to deferInitialization

sputh opened the pull request:
This is especially useful when you don't want to track any analytics or drop any analytic cookies prior to asking the users to opt in to analytics.
I've added a configuration (`deferInitialization`) and an API (`enableTracking`) to manage this.
If deferInitialization is set to true AND a customer has yet to opt in, we will locally store their pending logs until they either navigate away or opt in.

Note: This is a little gross because I'm checking whether or not to defer calling the function in EVERY public function. I considered using js decorators but it would require refactoring this to use classes. I considered iterating over an array of public methods and wrapping them, but that makes debugging much more difficult in the future. Ultimately, I opted for readability.

sputh commented on test/amplitude-client.js:
```diff
@@ -3331,4 +3331,69 @@ describe('setVersionName', function() {
+      });
+    });
+
+    describe('upon to opting into analytics', function () {
```
note: fix english 🤦🏻‍♀️

haoliu-amp commented:
Look good to me, but would leave approval decision to Dan or Krishna.

djih commented on src/amplitude-client.js:
```diff
@@ -1552,4 +1598,33 @@ if (BUILD_COMPAT_2_0) {
+ * @private
+ */
+AmplitudeClient.prototype._shouldDeferCall = function _shouldDeferCall() {
+  return this._pendingReadStorage || this._initializationDeferred;
```
what is pending read storage?

sputh replied:
we use this with `AsyncStorage` (react-native) to batch calls

sputh replied:
Actually, looking closer to the code, it's not only used for react-native
cc: @blazzy

blazzy replied:
It's only used for async storage yeah.

djih replied:
👍

djih commented on src/amplitude-client.js:
```diff
@@ -75,6 +75,12 @@ AmplitudeClient.prototype.init = function init(apiKey, opt_userId, opt_config, o
   }
 
+  var hasExistingCookie = !!this.cookieStorage.get(this.options.cookieName + this._storageSuffix);
+  if (opt_config && opt_config.deferInitialization && !hasExistingCookie) {
```
ok so let's say a user accepts and allows tracking and revisits the page the next day. There would already be an existing cookie and we would skip _deferInitialization. However, wouldn't `this._shouldDeferCall()` still return true?

sputh replied:
No `this._shouldDeferCall()` would return `false` (for the most part, there's a short period of time in the `init` function where `this._pendingReadStorage` returns true while getting started).

It depends on how they revisit the page. If the page never refreshed and they have the same instance of Amplitude, then the new events will not be sent --- I've added a test for this

If it's a new instance of amplitude (e.g refresh page, navigate to another page), then `hasExistingCookie` will be true and `_initializationDeferred ` will never be set to true, thus `this._shouldDeferCall()` will be false as long as `this._pendingReadStorage` is also false (most of the time)

sputh replied:
actually, after fixing my tests, i realized that you were right! gratzi!

sputh replied:
After adding tests, and further inspecting, i realize you're right. Because `_initializationDeferred` is never set to false AND `this._storageSuffix` isn't defined until the try block below, it's always assuming `_initializationDeferred` is true.

I've added a test to reflect it and additional assert statements for the existing tests

sputh commented on src/amplitude-client.js:
```diff
@@ -1552,4 +1598,33 @@ if (BUILD_COMPAT_2_0) {
+ */
+AmplitudeClient.prototype._deferInitialization = function _deferInitialization(apiKey, opt_userId, opt_config, opt_callback) {
+  this._initializationDeferred = true;
+  this._q.push(['init', apiKey, opt_userId, opt_config, opt_callback]);
```
we're never pushing init to `_q` and it's ultimately doing what we ne

sputh replied:
it doesn't seem like there's any potential risk to re-using it. I think it would add more confusion if we had multiple queues that are vaguely defined

djih replied:
👍

djih replied:
I hope this doesn't cause some weird circular queueing behavior where you defer a function that pushes to this or something, etc 🤔 (seems fine I guess?)

sputh replied:
i can't imagine it having circular dependency issues 🤔 
since it's a queue, it will run functions in order and it can tolerate if we pass in a function multiple times

djih commented on src/amplitude-client.js:
```diff
@@ -1552,4 +1598,33 @@ if (BUILD_COMPAT_2_0) {
+ * @public
+ */
+AmplitudeClient.prototype.enableTracking = function enableTracking() {
+  // This will call init (which drops the cookie) and will run any pending tasks
```
nvm figured it out

sputh replied:
i was intending to drop the cookie so that when init gets called again (bc it's in `this._q`), it will short circuit the `this._deferInitialization()` and not reset `this._initialization` to `true` (this doesn't really work in this commit, but is fixed in the one i just pushed bc `this._initializationDeferred` is explicitly set to `false` in `enableTracking()`, but _both_ are needed for it to work)

sputh commented on src/amplitude-client.js:
```diff
@@ -79,6 +79,12 @@ AmplitudeClient.prototype.init = function init(apiKey, opt_userId, opt_config, o
     this.options.apiKey = apiKey;
     this._storageSuffix = '_' + apiKey + this._legacyStorageSuffix;
 
+    var hasExistingCookie = !!this.cookieStorage.get(this.options.cookieName + this._storageSuffix);
```
moved into the try catch to ensure that `this._storageSuffix` is defined

djih replied:
i kind of like that you're using the cookie to signal if the user has accepted tracking

sputh replied:
i feel like that's the safest bet to carry over. i haven't fully thought of the case where our cookie expires tho.... 😬

djih commented on src/amplitude-client.js:
```diff
@@ -79,6 +79,12 @@ AmplitudeClient.prototype.init = function init(apiKey, opt_userId, opt_config, o
     this._storageSuffix = '_' + apiKey + this._legacyStorageSuffix;
 
+    var hasExistingCookie = !!this.cookieStorage.get(this.options.cookieName + this._storageSuffix);
+    if (opt_config && opt_config.deferInitialization && !hasExistingCookie) {
```
note for existing users, they won't get the option to defer tracking because they will already have a cookie, i guess we just need to communicate that to our customers and the people that requested this feature

sputh replied:
yes, this will only work on new users without existing cookies and they have to add the config flag

djih commented on src/amplitude-client.js:
```diff
@@ -766,6 +772,10 @@ AmplitudeClient.prototype.saveEvents = function saveEvents() {
  */
 AmplitudeClient.prototype.setDomain = function setDomain(domain) {
+  if (this._shouldDeferCall()) {
+    return this._q.push(['setDomain', domain]);
```
is it weird to return the result of array.push? or are you trying to shorthand
```javascript
{
   this._q.push...
   return
}
```

sputh replied:
shorthand. im doing this for the `set__`  where it's not expecting anything in return

sputh commented:
@djih for sanity's sake, I took your suggestion and added a test to ensure that existing users (e.g. user already has a cookie), will continue to send events even if `deferInitialization` is set to `true`

I've also updated the snippet and the minor version.
